File: src/utils/gui_wrapper.ts
/**@noSelf **/
export interface IGUIWrapper {
    button(node_name: string, callback: () => void, category?: number): DruidNode;
    block(node_name: string, is_block_enabled: boolean): void;
    block_category(category: number, is_block_enabled: boolean): void;
    is_bloked(node_name: string): boolean;
    set_befor_block_category_func(befor_block_category_: (category: number, node: string, is_blocked: boolean) => void): void;
    remove_category(category: number): void;
    remove(druid_node: DruidNode): void;
}

export function GUIWrapper(druid: DruidClass): IGUIWrapper {
    type ComponentDruid = {
        node: DruidNode,
        node_name: string,
        category: number
    };

    const components: ComponentDruid[] = [];
    const black_list: ComponentDruid[] = [];
    let befor_block_category: (category: number, node: string, is_blocked: boolean) => void | undefined;

    function button(node_name: string, callback: () => void, category = 0): DruidNode {
        const c = {
            node_name: node_name,
            node: druid.new_button(node_name, callback),
            category: category
        };
        components.push(c);
        return c.node;
    }

    function remove_category(category: number) {
        block_category(category, false);
        for (let i = components.length - 1 ; i > 0; i--) {
            const c = components[i];
            if(c.category == category) {
                pprint('remove');
                druid.remove(c.node);
            }
        }
    }

    function remove(druid_node: DruidNode) {
        const comp = components.findIndex(val => val.node == druid_node);
        if(comp!=-1) {
            block(components[comp].node_name, false);
            druid.remove(components[comp].node);
            components.splice(comp, 1);
        }
    }

    function set_befor_block_category_func(befor_block_category_: (category: number, node: string, is_blocked: boolean) => void) {
        befor_block_category = befor_block_category_;
    }
    function block(node_name: string, is_block_enabled: boolean): void {
        const node = components.find(comp => comp.node_name == node_name);
        if (node != null) {
            if (is_block_enabled) {
                if (black_list.find(n => n.node_name == node_name) != null) return;
                black_list.push(node);
            } else {
                const i = black_list.findIndex(v => v.node_name == node_name);
                if (i != -1)
                    black_list.splice(i, 1);
                else {
                    //pprint("node " + node.node_name + " not found");
                }

            }
            druid.set_blacklist(get_all_nodes(black_list));
        } else {
            //pprint('node ' + node_name + " not found");
        }

    }

    function block_category(category: number, is_block_enabled: boolean): void {
        const nodes = components.filter(comp => comp.category == category);
        if (nodes.length == 0) {
            pprint("category " + category + " empty!");
            return;
        }
        else if (nodes != null) {
            if (is_block_enabled) {
                nodes.forEach(node => {
                    if (black_list.find(n => n.node_name == node.node_name) == null) {
                        black_list.push(node);
                        if (befor_block_category != null)
                            befor_block_category(category, node.node_name, true);
                    }
                });
            } else {
                nodes.forEach(node => {
                    const i = black_list.findIndex(v => v.node_name == node.node_name);
                    if (i != -1) {
                        black_list.splice(i, 1);
                        if (befor_block_category != null)
                            befor_block_category(category, node.node_name, false);
                    }
                    else {
                        //pprint("node " + tostring(node.node_name) + " with category " + node.category + " not found");
                    }
                });

            }
            //pprint("block node category " + category + " " + is_block_enabled);
            druid.set_blacklist(get_all_nodes(black_list));
        }
    }

    function is_bloked(node_name: string): boolean {
        return black_list.find(n => n.node_name == node_name) != null;
    }
    function get_all_nodes(components: ComponentDruid[]): DruidNode[] {
        return components.map<DruidNode>(value => {
            return value.node;
        });
    }

    return { is_bloked, button, block, block_category, set_befor_block_category_func,remove_category,remove };
}

File: app/utils/gui_wrapper.lua
local ____lualib = require("lualib_bundle")
local __TS__ArrayFindIndex = ____lualib.__TS__ArrayFindIndex
local __TS__ArraySplice = ____lualib.__TS__ArraySplice
local __TS__ArrayFind = ____lualib.__TS__ArrayFind
local __TS__ArrayFilter = ____lualib.__TS__ArrayFilter
local __TS__ArrayForEach = ____lualib.__TS__ArrayForEach
local __TS__ArrayMap = ____lualib.__TS__ArrayMap
local ____exports = {}
function ____exports.GUIWrapper(druid)
    local block, block_category, get_all_nodes, components, black_list, befor_block_category
    function block(node_name, is_block_enabled)
        local node = __TS__ArrayFind(
            components,
            function(____, comp) return comp.node_name == node_name end
        )
        if node ~= nil then
            if is_block_enabled then
                if __TS__ArrayFind(
                    black_list,
                    function(____, n) return n.node_name == node_name end
                ) ~= nil then
                    return
                end
                black_list[#black_list + 1] = node
            else
                local i = __TS__ArrayFindIndex(
                    black_list,
                    function(____, v) return v.node_name == node_name end
                )
                if i ~= -1 then
                    __TS__ArraySplice(black_list, i, 1)
                else
                end
            end
            druid:set_blacklist(get_all_nodes(black_list))
        else
        end
    end
    function block_category(category, is_block_enabled)
        local nodes = __TS__ArrayFilter(
            components,
            function(____, comp) return comp.category == category end
        )
        if #nodes == 0 then
            pprint(("category " .. tostring(category)) .. " empty!")
            return
        elseif nodes ~= nil then
            if is_block_enabled then
                __TS__ArrayForEach(
                    nodes,
                    function(____, node)
                        if __TS__ArrayFind(
                            black_list,
                            function(____, n) return n.node_name == node.node_name end
                        ) == nil then
                            black_list[#black_list + 1] = node
                            if befor_block_category ~= nil then
                                befor_block_category(category, node.node_name, true)
                            end
                        end
                    end
                )
            else
                __TS__ArrayForEach(
                    nodes,
                    function(____, node)
                        local i = __TS__ArrayFindIndex(
                            black_list,
                            function(____, v) return v.node_name == node.node_name end
                        )
                        if i ~= -1 then
                            __TS__ArraySplice(black_list, i, 1)
                            if befor_block_category ~= nil then
                                befor_block_category(category, node.node_name, false)
                            end
                        else
                        end
                    end
                )
            end
            druid:set_blacklist(get_all_nodes(black_list))
        end
    end
    function get_all_nodes(components)
        return __TS__ArrayMap(
            components,
            function(____, value)
                return value.node
            end
        )
    end
    components = {}
    black_list = {}
    local function button(node_name, callback, category)
        if category == nil then
            category = 0
        end
        local c = {
            node_name = node_name,
            node = druid:new_button(node_name, callback),
            category = category
        }
        components[#components + 1] = c
        return c.node
    end
    local function remove_category(category)
        block_category(category, false)
        do
            local i = #components - 1
            while i > 0 do
                local c = components[i + 1]
                if c.category == category then
                    pprint("remove")
                    druid:remove(c.node)
                end
                i = i - 1
            end
        end
    end
    local function remove(druid_node)
        local comp = __TS__ArrayFindIndex(
            components,
            function(____, val) return val.node == druid_node end
        )
        if comp ~= -1 then
            block(components[comp + 1].node_name, false)
            druid:remove(components[comp + 1].node)
            __TS__ArraySplice(components, comp, 1)
        end
    end
    local function set_befor_block_category_func(befor_block_category_)
        befor_block_category = befor_block_category_
    end
    local function is_bloked(node_name)
        return __TS__ArrayFind(
            black_list,
            function(____, n) return n.node_name == node_name end
        ) ~= nil
    end
    return {
        is_bloked = is_bloked,
        button = button,
        block = block,
        block_category = block_category,
        set_befor_block_category_func = set_befor_block_category_func,
        remove_category = remove_category,
        remove = remove
    }
end
return ____exports
